### PR TITLE
Feat/tls

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:
tls
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: curp/src/client/mod.rs [hunk 1]
```diff
@@ -274,6 +296,8 @@ impl ClientBuilder {
             UnaryConfig::new(
                 *self.config.propose_timeout(),
                 *self.config.wait_synced_timeout(),
+                #[cfg(not(madsim))]
+                None,
```
Comment (iGxnon) [inline curp/src/client/mod.rs:300]: Why None?

File: xline-client/src/lib.rs [hunk 2]
```diff
@@ -403,8 +425,19 @@ impl ClientOptions {
...
+    }
+
+    /// Set `tls_config`
+    #[inline]
+    #[must_use]
+    #[cfg(not(madsim))]
+    pub fn with_tls_config(self, tls_config: ClientTlsConfig) -> Self {
+        Self {
+            tls_config: Some(tls_config),
+            ..self
```
Comment (iGxnon) [inline xline-client/src/lib.rs:440]: Should `ClientTlsConfig` be a part of `ClientConfig`?

File: xline/src/server/xline_server.rs [hunk 3]
```diff
@@ -71,6 +73,12 @@ pub struct XlineServer {
     compact_cfg: CompactConfig,
     /// Server timeout
     server_timeout: ServerTimeout,
+    /// Client tls config
+    #[cfg(not(madsim))]
+    client_tls_config: Option<ClientTlsConfig>,
```
Comment (iGxnon) [inline xline/src/server/xline_server.rs:78]: Same as above, should it be a client config?

File: Cargo.toml [hunk 4]
```diff
@@ -22,9 +22,9 @@ resolver = "2"
...
-madsim = { git = "https://github.com/madsim-rs/madsim.git", rev = "5eb4198", package = "madsim" }
-tokio = { git = "https://github.com/madsim-rs/madsim.git", rev = "5eb4198", package = "madsim-tokio" }
-tonic = { git = "https://github.com/madsim-rs/madsim.git", rev = "5eb4198", package = "madsim-tonic" }
-tonic-build = { git = "https://github.com/madsim-rs/madsim.git", rev = "5eb4198", package = "madsim-tonic-build" }
+# # TODO: We should remove it when the madsim updates its release in the crate.io
+# [patch.crates-io]
+# madsim = { git = "https://github.com/madsim-rs/madsim.git", rev = "5eb4198", package = "madsim" }
+# tokio = { git = "https://github.com/madsim-rs/madsim.git", rev = "5eb4198", package = "madsim-tokio" }
+# tonic = { git = "https://github.com/madsim-rs/madsim.git", rev = "5eb4198", package = "madsim-tonic" }
+# tonic-build = { git = "https://github.com/madsim-rs/madsim.git", rev = "5eb4198", package = "madsim-tonic-build" }
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply Cargo.toml:30]: Removed

File: curp/src/client/mod.rs (hunk 1 above)
Comment (themanforfree) [reply curp/src/client/mod.rs:300]: Fixed

File: xline-client/src/lib.rs (hunk 2 above)
Comment (themanforfree) [reply xline-client/src/lib.rs:440]: Currently, the content in `ClientConfig` is directly obtained from command line arguments or configuration files, while `ClientTlsConfig` needs to read the certificate file and cannot be obtained directly, they can't contain each other for now, but our config logic may be modified when config is refactored in the future

File: xline/src/server/xline_server.rs (hunk 3 above)
Comment (themanforfree) [reply xline/src/server/xline_server.rs:78]: Same as above

File: curp/src/members.rs [hunk 5]
```diff
@@ -107,6 +109,8 @@ impl ClusterInfo {
     }
 
     /// Construct a new `ClusterInfo` from `FetchClusterResponse`
+    /// # Panics
+    /// panic if `cluster.members` does contain `self_addr`
```
Comment (Phoenix500526) [inline curp/src/members.rs:113]: doesn't contain `self_addr`?

File: curp/src/rpc/connect.rs [hunk 6]
```diff
@@ -64,15 +66,16 @@ impl FromTonicChannel for InnerProtocolClient<Channel> {
 /// Connect to a server
 async fn connect_to<Client: FromTonicChannel>(
     id: ServerId,
-    mut addrs: Vec<String>,
+    addrs: Vec<String>,
+    #[cfg(not(madsim))] tls_config: Option<ClientTlsConfig>,
```
Comment (Phoenix500526) [inline curp/src/rpc/connect.rs:70]: We can define two `connect_to` methods, one with `#[cfg(not(madsim))]` while the other with `#[cfg(madsim)]`. And define a mock `ClientTlsConfig>` in `simulation/src/lib.rs`. Using conditional compile to pick which one should be use. Therefore, we can remove many `#[cfg(madsim)]` in other codes of this pr.

**Review by Phoenix500526 — COMMENTED** [review]
Etcd supports authenticate a user via certificate. In this way, we don't need the password any more. But I don't see any relevant change in this pr. Will this feature be implemented in other pr?  FYI: https://www.cnblogs.com/traditional/p/17444921.html

File: curp/src/members.rs (hunk 5 above)
Comment (themanforfree) [reply curp/src/members.rs]: Fixed

File: curp/src/rpc/connect.rs (hunk 6 above)
Comment (themanforfree) [reply curp/src/rpc/connect.rs]: Refactored

**Review by iGxnon — DISMISSED** [review]

File: utils/src/lib.rs [hunk 7]
```diff
@@ -143,6 +143,26 @@
...
+#[allow(missing_copy_implementations)] // Same as real config
+#[non_exhaustive]
+pub struct ClientTlsConfig;
+
+/// Fake `ServerTlsConfig` for `madsim`
+#[cfg(madsim)]
+#[derive(Debug, Clone)]
+#[allow(missing_copy_implementations)] // Same as real config
+#[non_exhaustive]
+pub struct ServerTlsConfig;
```
Comment (Phoenix500526) [inline utils/src/lib.rs:164]: Is it better to place these two struct in `simulation/src/lib.rs`?

**Review by Phoenix500526 — COMMENTED** [review]
I think that hardcode the `*.crt`, `*.key` in crate is not a good practice. I recommend to place them in the top directory of Xline, and named it `fixtures`, or something like that.

File: utils/src/lib.rs (hunk 7 above)
Comment (themanforfree) [reply utils/src/lib.rs:164]: `xline` also depends on these fake configs, and `xline` shouldn't depend on `simulation`

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by iGxnon — APPROVED** [review]
LGTM

**Review by iGxnon — APPROVED** [review]

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**